runtime/io/common/src/aws/smithy/kotlin/runtime/io/CompletingByteReadChannel.kt

```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package aws.smithy.kotlin.runtime.io
+
+import aws.smithy.kotlin.runtime.util.InternalApi
+import aws.smithy.kotlin.runtime.util.encodeBase64String
+import kotlinx.coroutines.CompletableDeferred
+
+/**
+ * A channel which uses the digest of an underlying [HashingByteReadChannel] to complete a [CompletableDeferred]
+ * @param deferredChecksum The [CompletableDeferred] to be completed when the underlying channel is exhausted
+ * @param hashingChannel the [HashingByteReadChannel] which will be digested and used to complete [deferredChecksum]
+ */
+@InternalApi
+public class CompletingByteReadChannel(
+    private val deferredChecksum: CompletableDeferred<String>,
+    private val hashingChannel: HashingByteReadChannel,
+) : SdkByteReadChannel by hashingChannel {
```

Comment:
I agree, I thought about how to make it more generic. The `CompletingByteReadChannel` could take a function as input which would be executed to obtain the value after the underlying body is exhausted.

```kotlin
public class CompletingByteReadChannel<T>(
    private val deferred: CompletableDeferred<T>,
    private val channel: SdkByteReadChannel,
    private val block: suspend () -> (T),
) : SdkByteReadChannel by channel {
        public override suspend fun read(sink: SdkBuffer, limit: Long): Long =
        channel.read(sink, limit).also {
            if (channel.isClosedForRead) {
                deferredChecksum.complete(block())
            }
        }
}
```

Then, it can be used for hashing channels like this:

```kotlin
val deferredChecksum = CompletableDeferred()
val channel = HashingByteReadChannel(...)
val completingChannel = CompletingByteReadChannel<String>(deferredChecksum, channel) { channel.digest().encodeBase64String() }
```

What do you think about this?

Reply:
That definitely seems like a cleaner separation of concerns (although I'm now noticing neither the original nor this have error handling and we'd probably want some). We could probably even add a `HashingByteReadChannel.asCompleting()` extension function to handle the extraction/encoding of the final digest.

Will we ever need to hash a channel without completing a `Deferred`? Are there other known use cases for a completing channel besides hashing?

Reply:
**comment**: I'm not sure I would even expose this from `io`. I'd probably just err on the side of making it a small utility as part of the implementation (`internal` to signing/chunk implementation). I don't think we have any other use cases for this in the runtime and this is very simple wrapper code.

Reply:
I refactored this by instead adding an optional `CompletableDeferred` parameter to the `HashingSource`/`HashingByteReadChannel`.

If it is non-null, it will be completed using the `digest()` of the hash function once the underlying data source is exhausted.